### PR TITLE
#159557075 An endpoint to delete a role

### PR DESCRIPTION
**Pull request by nzediegwu1**
#### What does this PR do?
Allows a user to delete a role

#### Description of Task to be completed?
As a developer, I should be able to delete a role  by making a request to `/roles/<role_id>`

#### Acceptance Criteria:
**GIVEN** I have a valid authentication token
**WHEN**  I send a `DELETE` request to the `/roles/<role_id>` endpoint
**THEN** I will receive confirmation of the deletion of a role

#### How should this be manually tested?
- pull the branch: `git fetch origin ft-delete-role-159557075:ft-delete-role-159557075`
- checkout to the branch: `git checkout ft-delete-role-159557075`
- setup using commands: `pipenv shell`, `pipenv install --dev`
- open roles table in database and create new sample roles to be deleted
- run the app: `flask run`
- on postman: set bearer token and make a DELETE request to `/roles/<role_id>` where role_id is an existing role, to get a 200 response upon success
- make another DELETE request to the `/roles/<role_id>` route using a non-existing role_id, to get a 404 response
- remove bearer token and make another get request to get a 401 error response

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#159557075](https://www.pivotaltracker.com/story/show/159557075)

#### What pivotal tracker stories does this block?
None

#### What pivotal tracker stories are blocking this?
1. [An endpoint to create a new role](https://www.pivotaltracker.com/story/show/159557028)

### Postman Documentation
[Run in Postman](https://app.getpostman.com/run-collection/16b6480fed8b2d7b3037)

#### Screenshots (if appropriate)
screen shot 2018-08-07 at 4 00 51 pm
screen shot 2018-08-07 at 4 01 27 pm
screen shot 2018-08-07 at 4 02 25 pm

### CODE REVIEW

**Comment by Victoradukwu** [comment]
@nzediegwu1 Please edit your commit message to reflect the team's convention.

File: tests/conftest.py [hunk 1]
```diff
@@ -48,78 +48,83 @@ def client(app):
...
+        Role(title="Operations Manager"),
     ]
     return roles
 
 
-@pytest.fixture(scope='module')
+@pytest.fixture(scope="module")
 def new_role(app):
-    role = Role(title='Operations Intern')
+    role = Role(title="Operations Intern")
```
Comment (purpose50) [inline tests/conftest.py]: Hi, @kenoseni , I can see you are trying to change the quotes from single to double quotes, actually, Pep8 doesn't care what kind of quotes you use, the convention is to pick one and stick to it, the team has basically been using a mix of both which is not good. But I think changing it here makes it difficult for me to know which changes entail your exact task. I suggest you remove them so I can see what exactly is your task. we can always create a story to fix the single/double quotes issue when we agree one.

File: tests/conftest.py [hunk 2]
```diff
@@ -48,78 +48,83 @@ def client(app):
...
 
-@pytest.fixture(scope='module')
+
+@pytest.fixture(scope="module")
+def test_role(app):
+    new_role = Role(title="Software developer")
+    return new_role
+
+@pytest.fixture(scope="module")
 def new_role(app):
```
Comment (GbengaOshinaga) [inline tests/conftest.py]: @nzediegwu1 This fixture is repeated below, delete one of them

**Comment by nzediegwu1** [comment]
@Victoradukwu, thanks for the feedback. Most Implemented. I included the `flask db upgrade` command because there is a possibility that the local repo of the developer is not up to date with the model required to test this feature.

File: tests/conftest.py (hunk 2 above)
Comment (nzediegwu1) [reply tests/conftest.py]: Thanks, will do

File: tests/conftest.py (hunk 1 above)
Comment (nzediegwu1) [reply tests/conftest.py]: Thanks for the feedback. Actually, the changes were made by my editors code auto-formatter whenever I save. Will try to fix that

**Comment by nzediegwu1** [comment]
Thanks @kenoseni, corrected!

File: tests/conftest.py (hunk 2 above)
Comment (nzediegwu1) [reply tests/conftest.py]: Done!

File: tests/conftest.py (hunk 1 above)
Comment (nzediegwu1) [reply tests/conftest.py]: Done!

**Comment by Victoradukwu** [comment]
@nzediegwu1 Please resolve the merge conflicts on the branch.

File: api/views/role.py [hunk 3]
```diff
@@ -20,15 +25,48 @@ def get(self):
...
+
+        Returns:
+            (dict) -- Returns status and success message
+        """
+        role = Role.get_or_404(role_id)
+        if not role:
+            raise ValidationError(
+                {
+                    'message': serialization_errors['not_found'].format('Role')
+                }, 404)
```
Comment (Victoradukwu) [inline api/views/role.py]: @nzediegwu1 The block of code in lines 61-65 is no longer necessary. The `get_or_404` method has already taken care of that validation.

File: tests/test_role_endpoints.py [hunk 4]
```diff
@@ -23,3 +28,51 @@ def test_get_role_endpoint(self, client, init_db, auth_header, new_roles):
...
+
+    def test_delete_role_without_token(self, client, init_db):
+        """
+        Tests that 401 is returned when token is not provided
+        """
+        response = client.delete(f'{BASE_URL}/roles/-LX%%%tghrfe5')
+        response_json = json.loads(response.data.decode(CHARSET))
+        assert response.status_code == 401
+        assert response_json['status'] == 'error'
+        assert response_json['message'] == jwt_errors['NO_TOKEN_MSG']
```
Comment (Victoradukwu) [inline tests/test_role_endpoints.py]: @nzediegwu1 Please add a test for a case of deleting a `role` with an existing `user`.

File: api/views/role.py (hunk 3 above)
Comment (nzediegwu1) [reply api/views/role.py]: Okay, will do

File: tests/test_role_endpoints.py (hunk 4 above)
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: Done, however I cannot resolve merge conflict for now because the develop is failing

File: tests/conftest.py [hunk 5]
```diff
@@ -201,7 +207,7 @@ def auth_header(generate_token=generate_token):
     return {
         'Authorization': generate_token(),
         'Content-Type': MIMETYPE,
-        'Accept': MIMETYPE
+        'Accept': MIMETYPE,
```
Comment (GbengaOshinaga) [inline tests/conftest.py]: @nzediegwu1 Why did you add the trailing comma?

File: api/views/role.py (hunk 3 above)
Comment (nzediegwu1) [reply api/views/role.py]: done

File: tests/test_role_endpoints.py (hunk 4 above)
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: done

File: api/views/role.py (hunk 3 above)
Comment (GbengaOshinaga) [reply api/views/role.py]: @nzediegwu1 It doesn't seem like you've implemented @Victoradukwu's feedback

File: tests/test_role_endpoints.py (hunk 4 above)
Comment (GbengaOshinaga) [reply tests/test_role_endpoints.py]: @nzediegwu1 You haven't implemented this one also

Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: Sorry, my changes were discarded mysteriously after my previous rebase. I have reverted it back. You can check again

**Comment by nzediegwu1** [comment]
@GbengaOshinaga, sorry, my changes were discarded mysteriously after a previous rebase. I have reverted it back. You can check again

File: tests/conftest.py [hunk 6]
```diff
@@ -68,19 +67,14 @@ def new_role(app):
...
-    roles = [
-        Role(title='Operations Intern'),
-        Role(title='Operations Coordinator'),
-        Role(title='Operations Assistant'),
-        Role(title='Operations Manager'),
-    ]
-    return roles
+def test_role(app):
+    role = Role(title='Software developer')
+    return role
```
Comment (kenoseni) [inline tests/conftest.py]: Good job @nzediegwu1 . Presently the task to `create a new role` will be a blocker for your task because creating a role needs a `description` column and presently you only have a `title` column in your `Role model`. Your tests will fail once the `create new role task` is merged because there is no description field in the `role` you created. To avoid this I suggest the `create user role` should be merged so that you can update accordingly.

Comment (nzediegwu1) [reply tests/conftest.py]: Okay, will do

Comment (ibutiti) [reply tests/conftest.py]: This has been merged... You should refactor to use their mocks to save code.

File: tests/test_role_endpoints.py (hunk 4 above)
Comment (ibutiti) [reply tests/test_role_endpoints.py]: Please add a test for deleting a role with a user that's been deleted.

**Comment by nzediegwu1** [comment]
@ibutiti, Please explain your requested changes on my PR: `Please add a test for deleting a role with a user that's been deleted.`,

**Comment by nzediegwu1** [comment]
@ibutiti, Please explain: `You should refactor to use their mocks to save code.`

File: tests/conftest.py (hunk 6 above)
Comment (nzediegwu1) [reply tests/conftest.py]: @ibutiti, Please explain: `You should refactor to use their mocks to save code.`

**Comment by ibutiti** [comment]
I mean, add a user to the role, soft delete the user then delete the role to cover that edge case.

File: tests/test_role_endpoints.py (hunk 4 above)
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: okay

File: tests/conftest.py (hunk 6 above)
Comment (ibutiti) [reply tests/conftest.py]: Hey @nzediegwu1, I mean using the already existing mocks/fixtures for adding single role in the db instead of creating your own as you've done here.

sorry, just seen this response.

File: tests/test_role_endpoints.py (hunk 4 above)
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: @ibutiti Done!

File: tests/test_role_endpoints.py [hunk 7]
```diff
@@ -167,3 +169,93 @@ def test_create_role_fails_with_invalid_description(
...
+        new_user_with_role = new_test_user(user_data)
+        new_user_with_role.save()  # create new user
+        response = client.delete(
+            f'{BASE_URL}/roles/{role_id}', headers=auth_header)
+        response_json = json.loads(response.data.decode(CHARSET))
+        assert response.status_code == 403
+        assert response_json['status'] == 'error'
+        assert response_json['message'] == database_errors[
+            'model_delete_children'].format('Role', 'User(s)')
+        """
```
[the remark this answers is not in the document]
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: I added it to the existing user test because I feel it is best to delete that same role created which previously was not deletable in other to show the contrasting behaviour.

[the remark this answers is not in the document]
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: That would mean creating a new center, new role, and then new user having the center_id of the new center created and role_id of the new role created, all of which are meant to be unique and hence cannot be abstracted without creating new data

[the remark this answers is not in the document]
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: I did it this way because I feel creating a new role, no user, deleting the new user and then deleting the new role before making assertions without any assertions in between to confirm that actions performed were successful would make the test to be not verifiable

[the remark this answers is not in the document]
Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: I did it this way because I feel creating a new role, new user, deleting the new user and then deleting the new role before making assertions without any assertions in between to confirm that actions performed were successful would make the test to be not verifiable

[the remark this answers is not in the document]
Comment (ibutiti) [reply tests/test_role_endpoints.py]: Like you said, `existing_user=true` and `existing_user=false` are two different scenarios. They should be different tests.

It is possible to abstract the fixture and make the two entries unique. You can use the random string generator I created to generate random names etc for each of the unique fields. Have a look at my implementation.

**Comment by nzediegwu1** [comment]
@ibutiti, feedback implemented

File: api/views/role.py [hunk 8]
```diff
@@ -4,10 +4,13 @@
 from flask import jsonify, request
 
 from main import api
+
```
Comment (SEUNAGBEYE) [inline api/views/role.py]: @nzediegwu1 Please remove this space.

File: tests/conftest.py [hunk 9]
```diff
@@ -174,6 +210,18 @@ def new_center(app):
     return center
 
 
+@pytest.fixture(scope='module')
+def new_test_center(app):
+    params = {
+        'name': fake.name(),
```
Comment (SEUNAGBEYE) [inline tests/conftest.py]: nice

File: tests/conftest.py [hunk 10]
```diff
@@ -77,6 +79,12 @@ def new_role(app):
     return role
 
 
+@pytest.fixture(scope='module')
+def new_test_role(app):
```
Comment (appcypher) [inline tests/conftest.py]: There is an existing `new_role` fixture you can use for this.

File: tests/conftest.py [hunk 11]
```diff
@@ -133,6 +150,24 @@ def test_single_asset_category(app):
...
+    user_data = {  # new user-data having new center id and new role id
+        'id': f'{format(randint(0, 10))}',
+        'name': fake.first_name(),
+        'email': fake.email(),
+        'image_url': fake.image_url(),
+        'role_id': role_id,
+        'center_id': center_id
+    }
+    new_user_with_role = new_test_user(user_data)
+    return [new_user_with_role, role_id]  # create new user
```
Comment (SEUNAGBEYE) [inline tests/conftest.py]: Could you please return a tuple here.

File: tests/test_role_endpoints.py [hunk 12]
```diff
@@ -167,3 +169,83 @@ def test_create_role_fails_with_invalid_description(
...
+        assert response_json['status'] == 'error'
+        assert response_json['message'] == jwt_errors['NO_TOKEN_MSG']
+
+    def test_delete_role_with_existing_user(
+            self, client, init_db, user_with_role, request_ctx,
+            mock_request_obj_decoded_token, auth_header):
+        """
+        Tests that a role will not delete when role has existing user assigned
+        """
+        user_with_role[0].save()  # create new user in new role
```
Comment (SEUNAGBEYE) [inline tests/test_role_endpoints.py]: @nzediegwu1 Then you can unpack the tuple here: 
```
user, role_id = user_with_role
user.save()
response = client.delete( f'{BASE_URL}/roles/{role_id}, headers=auth_header)
```

File: tests/test_role_endpoints.py [hunk 13]
```diff
@@ -167,3 +169,83 @@ def test_create_role_fails_with_invalid_description(
...
+        assert response_json['message'] == database_errors[
+            'model_delete_children'].format('Role', 'User(s)')
+
+    def test_delete_role_after_user_deleted(
+            self, client, init_db, user_with_role, request_ctx,
+            mock_request_obj_decoded_token, auth_header):
+        """
+        Test for deleting a role after existing user has been deleted
+        """
+        user_with_role[0].save()  # create new user in new role
```
Comment (SEUNAGBEYE) [inline tests/test_role_endpoints.py]: Please use `tuple` unpacking here.

Comment (nzediegwu1) [reply tests/test_role_endpoints.py]: Done

File: tests/conftest.py (hunk 10 above)
Comment (nzediegwu1) [reply tests/conftest.py]: The new_role fixture is being used somewhere else. Since new role requires a unique title, it cannot be reused
